src/platforms/esp32/components/avm_builtins/i2c_resource.c

```diff
@@ -0,0 +1,743 @@
+    if (UNLIKELY(!to_i2c_resource(i2c_resource, &rsrc_obj, ctx))) {
+        ESP_LOGE(TAG, "Failed to convert i2c_resource");
+        RAISE_ERROR(BADARG_ATOM);
+    }
+
+    //
+    // ensure there is not a currently running transmission
+    //
+    if (UNLIKELY(!term_is_invalid_term(rsrc_obj->transmitting_pid))) {
+        ESP_LOGE(TAG, "nif_read_bytes: Another process is already transmitting");
+
+        // {error, {einprogress, Pid :: pid()}}
+        if (UNLIKELY(memory_ensure_free(ctx, 2 * TUPLE_SIZE(2)) != MEMORY_GC_OK)) {
+            ESP_LOGW(TAG, "Failed to allocate memory: %s:%i.\n", __FILE__, __LINE__);
+            return OUT_OF_MEMORY_ATOM;
+        }
+        term reason = create_pair(
+            ctx,
+            globalcontext_make_atom(global, EINPROGRESS_ATOMSTR),
+            rsrc_obj->transmitting_pid
+        );
+        return create_error_tuple(ctx, reason);
+    }
+
+    //
+    // extract the arguments
+    //
+    term req = argv[1];
+    int arity = term_get_tuple_arity(req);
+
+    term address = term_get_tuple_element(req, 0);
+    uint8_t addr = term_to_int32(address);
+
+    term read_bytes = term_get_tuple_element(req, 1);
+    avm_int_t read_count = term_to_int32(read_bytes);
+
+    term register_ = term_invalid_term();
+    uint8_t register_address = 0;
+    if (arity == 3) {
+        register_ = term_get_tuple_element(req, 2);
+        register_address = term_to_int32(register_);
+    }
+
+    //
+    // initialize the output buffer
+    //
+
+    // {ok, Data :: binary()}
+    if (UNLIKELY(memory_ensure_free_opt(ctx, TUPLE_SIZE(2) + term_binary_heap_size(read_count), MEMORY_NO_GC) != MEMORY_GC_OK)) {
+        ESP_LOGW(TAG, "Failed to allocate memory: %s:%i.\n", __FILE__, __LINE__);
+        return OUT_OF_MEMORY_ATOM;
+    }
+    term data = term_create_uninitialized_binary(read_count, &ctx->heap, ctx->global);
+    uint8_t *buf = (uint8_t *) term_binary_data(data);
+
+    //
+    // Enqueue and run the I2C commands
+    //
+
+    esp_err_t err;
+    i2c_cmd_handle_t cmd = i2c_cmd_link_create();
+    {
+        err = i2c_master_start(cmd);
+        CHECK_ERROR(ctx, err, "nif_read_bytes; enqueue start bit");
+
+        if (!term_is_invalid_term(register_)) {
+            err = i2c_master_write_byte(cmd, (addr << 1) | I2C_MASTER_WRITE, 0x1);
+            err = i2c_master_write_byte(cmd, register_address, 0x1);
+            err = i2c_master_start(cmd);
+        }
+
+        err = i2c_master_write_byte(cmd, (addr << 1) | I2C_MASTER_READ, true);
+        CHECK_ERROR(ctx, err, "nif_read_bytes; enqueue I2C bus address");
+
+        err = i2c_master_read(cmd, buf, read_count, I2C_MASTER_LAST_NACK);
+        CHECK_ERROR(ctx, err, "nif_read_bytes; enqueue write data");
+
+        err = i2c_master_stop(cmd);
+        CHECK_ERROR(ctx, err, "nif_write_bytes; enqueue stop bit");
+    }
+    // TODO allow customization of timeout
+    err = i2c_master_cmd_begin(rsrc_obj->i2c_num, cmd, portMAX_DELAY);
```

Comment:
We should avoid introducing more blocking functions inside new drivers.

Reply:
There isn't really an option, since there is no async API for I2C.  I'd rather we just mark this as an issue and move on.  It is clearly also an issue for the existing port, so what's sauce for the goose is sauce for the gander.

Reply:
I have added support for timeouts on i2c messages (default 500ms, but configurable in `i2c:open/1` params).

The only alternative I can see for setting up non-blocking calls on i2c APIs is to create a separate RTOS task, with queues between the task and VM.  This is a non-trivial amount of work, and would need to be done for the port-based implementation, as well.  It seems pretty far out of scope for the intent of this PR, which is to provide similar functionality as the port-based implementation, but using Nif-based resources, for a performance improvement and lower memory overhead.